File: src/PokemonBattle.EffectsP/Items/PowerModifier.py
class TypeItem(ItemE):
    def __new__(cls, id, type):
        return ItemE.__new__(cls, id)
    def __init__(self, id, type):
        self.Type = type
    def PowerModifier(self, atk):
        if atk.Type == self.Type:
            return 0x1333
        return 0x1000
I(TypeItem(11, BattleType.Bug)) #silverpowder
I(TypeItem(17, BattleType.Steel)) #metal coat
I(TypeItem(20, BattleType.Ground)) #soft sand
I(TypeItem(21, BattleType.Rock)) #hard stone
I(TypeItem(22, BattleType.Grass)) #miracle seed
I(TypeItem(23, BattleType.Dark)) #blackglasses
I(TypeItem(24, BattleType.Fighting)) #black belt
I(TypeItem(25, BattleType.Electric)) #magnet
I(TypeItem(26, BattleType.Water)) #mystic water
I(TypeItem(27, BattleType.Flying)) #sharp beak
I(TypeItem(28, BattleType.Poison)) #poison barb
I(TypeItem(29, BattleType.Ice)) #nevermeltice
I(TypeItem(30, BattleType.Ghost)) #spell tag
I(TypeItem(31, BattleType.Psychic)) #twistedspoon
I(TypeItem(32, BattleType.Fire)) #charcoal
I(TypeItem(33, BattleType.Dragon)) #dragon fang
I(TypeItem(34, BattleType.Normal)) #silk scarf
I(TypeItem(36, BattleType.Water)) #sea incense
I(TypeItem(75, BattleType.Fire)) #flame plate
I(TypeItem(76, BattleType.Water)) #splash plate
I(TypeItem(77, BattleType.Electric)) #zap plate
I(TypeItem(78, BattleType.Grass)) #meadow plate
I(TypeItem(79, BattleType.Ice)) #icicle plate
I(TypeItem(80, BattleType.Fighting)) #fist plate
I(TypeItem(81, BattleType.Poison)) #toxic plate
I(TypeItem(82, BattleType.Ground)) #earth plate
I(TypeItem(83, BattleType.Flying)) #sky plate
I(TypeItem(84, BattleType.Psychic)) #mind plate
I(TypeItem(85, BattleType.Bug)) #insect plate
I(TypeItem(86, BattleType.Rock)) #stone plate
I(TypeItem(87, BattleType.Ghost)) #spooky plate
I(TypeItem(88, BattleType.Dragon)) #draco plate
I(TypeItem(89, BattleType.Dark)) #dread plate
I(TypeItem(90, BattleType.Steel)) #iron plate
I(TypeItem(91, BattleType.Psychic)) #odd incense
I(TypeItem(92, BattleType.Rock)) #rock incense
I(TypeItem(94, BattleType.Water)) #wave incense
I(TypeItem(95, BattleType.Grass)) #rose incense

class Orb(ItemE):
    def __new__(cls, id, pm, type):
        return ItemE.__new__(cls, id)
    def __init__(self, id, pm, type):
        self.Pm = pm
        self.Type = type
    def PowerModifier(self, atk):
        if atk.Attacker.Pokemon.Forme.Type.Number == self.Pm and (atk.Type == BattleType.Dragon or atk.Type == self.Type):
            return 0x1333
        return 0x1000
I(Orb(1, 487, BattleType.Ghost))
I(Orb(2, 483, BattleType.Steel))
I(Orb(3, 484, BattleType.Water))

class MuscleBand(ItemE):
    def __new__(cls, id, category):
        return ItemE.__new__(cls, id)
    def __init__(self, id, category):
        self.Category = category
    def PowerModifier(self, atk):
        if atk.Move.Category == self.Category:
            return 0x1199
        return 0x1199
I(MuscleBand(43, MoveCategory.Physical))
I(MuscleBand(44, MoveCategory.Special)) #wise glasses

File: trunk/src/PBO.Game.Host/Triggers/PowerModifier.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using PokemonBattleOnline.Game;

namespace PokemonBattleOnline.Game.Host.Triggers
{
  internal static class PowerModifier
  {
    public static Modifier Execute(DefContext def)
    {
      var der = def.Defender;
      var atk = def.AtkContext;
      var aer = atk.Attacker;
      var c = def.Defender.Controller;

      Modifier m = Abilities(def);

      m *= AttackerItem(atk);

      m *= Move(def);
      //If user used Charge the previous turn and move is Electric type.
      if (atk.Type == BattleType.Electric && atk.Attacker.OnboardPokemon.GetCondition<int>("Charge") == c.TurnNumber) m *= 0x2000;
      //If user has been the target of Helping Hand this turn.
      #warning HelpingHand
      //If Water Sport was used by any Pokémon still on the field and move is Fire type.
      //If Mud Sport was used by any Pokémon still on the field and move is Electric type.
      if ((atk.Type == BattleType.Fire && c.Board.HasCondition("WaterSport")) ||
        (atk.Type == BattleType.Electric && c.Board.HasCondition("MudSport")))
        m *= 0x800;

      if ((atk.Type == BattleType.Electric && c.Board.HasCondition("ElectricTerrain") || atk.Type == BattleType.Grass && c.Board.HasCondition("GrassyTerrain")) && HasEffect.IsGroundAffectable(aer, true, false)) m *= 0x1800;
      else if (atk.Type == BattleType.Dragon && c.Board.HasCondition("MistyTerrain") && HasEffect.IsGroundAffectable(der, true, false)) m *= 0x800;

      return m;
    }
    private static Modifier Abilities(DefContext def)
    {
      var der = def.Defender;
      var atk = def.AtkContext;
      var move = atk.Move;
      var aer = atk.Attacker;
      var type = atk.Type;

      Modifier m = 0x1000;

      if (atk.HasCondition("Sukin")) m *= 0x14cd;
      else
        switch (aer.Ability)
        {
          case As.RIVALRY:
            {
              var u = aer.OnboardPokemon.Gender;
              var t = der.OnboardPokemon.Gender;
              if (u != PokemonGender.None && t != PokemonGender.None)
                m *= (Modifier)(u == t ? 0x1400 : 0xc00);
            }
            break;
          case As.IRON_FIST:
            if (move.Flags.IsFist) m *= 0x1333;
            break;
          case As.TECHNICIAN:
            if (def.BasePower <= 60) m *= 0x1800;
            break;
          case As.RECKLESS:
            if (move.HurtPercentage < 0 || move.Id == Ms.JUMP_KICK || move.Id == Ms.HI_JUMP_KICK) m *= 0x1333;
            break;
          case As.TOXIC_BOOST:
            if ((aer.State == PokemonState.PSN || aer.State == PokemonState.BadlyPSN) && move.Category == MoveCategory.Physical) m *= 0x1800;
            break;
          case As.FLARE_BOOST:
            if (move.Category == MoveCategory.Special && aer.State == PokemonState.BRN) m *= 0x1800;
            break;
          case As.ANALYTIC:
            {
              var turn = aer.LastMoveTurn;
              if (aer.Controller.ActingPokemons.All((p) => p.LastMoveTurn == turn)) m *= 0x14cd;
            }
            break;
          case As.SHEER_FORCE:
            if (atk.Move.HasProbabilitiedAdditonalEffects()) m *= 0x14cd;
            break;
          case As.SAND_FORCE:
            if ((type == BattleType.Rock || type == BattleType.Ground || type == BattleType.Steel) && aer.Controller.Weather == Weather.Sandstorm) m *= 0x14cd;
            break;
          case As.STRONG_JAW:
            if (move.StrongJaw()) m *= 0x1800;
            break;
          case As.TOUGH_CLAWS:
            if (move.Flags.NeedTouch) m *= 0x14cd;
            break;
          case As.MEGA_LAUNCHER:
            if (move.MegaLaucher()) m *= 0x1800;
            break;
        }

      //如果防御方是耐热特性，攻击方火属性技能威力×0.5。
      //如果防御方是干燥肌肤特性，攻击方火属性技能威力×1.25。 
      if (type == BattleType.Fire)
      {
        int id = def.Ability;
        if (id == As.HEATPROOF) m *= 0x800;
        else if (id == As.DRY_SKIN) m *= 0x1800;
      }
      
      if (type == BattleType.Dark || type == BattleType.Fairy)
      {
        var a = type == BattleType.Dark ? As.DARK_AURA : As.FAIRY_AURA;
        if (aer.Controller.OnboardPokemons.HasAbility(a) != null) m *= (Modifier)(aer.Controller.OnboardPokemons.HasAbility(As.AURA_BREAK) == null ? 0x1555 : 0xc00);
      }

      return m;
    }
    
    private static Modifier AttackerItem(AtkContext atk)
    {
      if (atk.HasCondition("Gem")) return 0x1800;
      switch (atk.Attacker.Item)
      {
        case Is.GRISEOUS_ORB:
          return Orb(atk, 487, BattleType.Ghost);
        case Is.ADAMANT_ORB:
          return Orb(atk, 483, BattleType.Steel);
        case Is.LUSTROUS_ORB:
          return Orb(atk, 484, BattleType.Water);
        case Is.SILVERPOWDER:
        case Is.INSECT_PLATE:
          return TypeItem(atk, BattleType.Bug);
        case Is.METAL_COAT:
        case Is.IRON_PLATE:
          return TypeItem(atk, BattleType.Steel);
        case Is.SOFT_SAND:
        case Is.EARTH_PLATE:
          return TypeItem(atk, BattleType.Ground);
        case Is.HARD_STONE:
        case Is.STONE_PLATE:
        case Is.ROCK_INCENSE:
          return TypeItem(atk, BattleType.Rock);
        case Is.MIRACLE_SEED:
        case Is.MEADOW_PLATE:
        case Is.ROSE_INCENSE:
          return TypeItem(atk, BattleType.Grass);
        case Is.BLACKGLASSES:
        case Is.DREAD_PLATE:
          return TypeItem(atk, BattleType.Dark);
        case Is.BLACK_BELT:
        case Is.FIST_PLATE:
          return TypeItem(atk, BattleType.Fighting);
        case Is.MAGNET:
        case Is.ZAP_PLATE:
          return TypeItem(atk, BattleType.Electric);
        case Is.MYSTIC_WATER:
        case Is.SPLASH_PLATE:
        case Is.WAVE_INCENSE:
        case Is.SEA_INCENSE:
          return TypeItem(atk, BattleType.Water);
        case Is.SHARP_BEAK:
        case Is.SKY_PLATE:
          return TypeItem(atk, BattleType.Flying);
        case Is.POISON_BARB:
        case Is.TOXIC_PLATE:
          return TypeItem(atk, BattleType.Poison);
        case Is.NEVERMELTICE:
        case Is.ICICLE_PLATE:
          return TypeItem(atk, BattleType.Ice);
        case Is.SPELL_TAG:
        case Is.SPOOKY_PLATE:
          return TypeItem(atk, BattleType.Ghost);
        case Is.TWISTEDSPOON:
        case Is.MIND_PLATE:
        case Is.ODD_INCENSE:
          return TypeItem(atk, BattleType.Psychic);
        case Is.CHARCOAL:
        case Is.FLAME_PLATE:
          return TypeItem(atk, BattleType.Fire);
        case Is.DRAGON_FANG:
        case Is.DRACO_PLATE:
          return TypeItem(atk, BattleType.Dragon);
        case Is.PIXIE_PLATE:
          return TypeItem(atk, BattleType.Fairy);
        case Is.SILK_SCARF:
          return TypeItem(atk, BattleType.Normal);
        case Is.MUSCLE_BAND:
          return Category(atk, MoveCategory.Physical);
        case Is.WISE_GLASSES:
          return Category(atk, MoveCategory.Special);
        default:
          return 0x1000;
      }
    }
    private static Modifier TypeItem(AtkContext atk, BattleType type)
    {
      return (Modifier)(atk.Type == type ? 0x1333 : 0x1000);
    }
    private static Modifier Orb(AtkContext atk, int pm, BattleType type)
    {
      return (Modifier)(atk.Attacker.Pokemon.Form.Species.Number == pm && (atk.Type == BattleType.Dragon || atk.Type == type) ? 0x1333 : 0x1000);
    }
    private static Modifier Category(AtkContext atk, MoveCategory category)
    {
      return (Modifier)(atk.Move.Category == category ? 0x1199 : 0x1000);
    }
    
    private static Modifier Move(DefContext def)
    {
      var der = def.Defender;
      var atk = def.AtkContext;
      Modifier m = 0x1000;

      switch (atk.Move.Id)
      {
        case Ms.FUSION_FLARE: //558
          {
            var b = der.Controller.Board;
            b.SetTurnCondition("FusionFlare");
            var o = b.GetCondition("LastMove");
            if (o != null && o.Move.Id == Ms.FUSION_BOLT && b.HasCondition("FusionBolt")) m = 0x2000;
          }
          break;
        case Ms.FUSION_BOLT: //559
          {
            var b = der.Controller.Board;
            b.SetTurnCondition("FusionBolt");
            var o = b.GetCondition("LastMove");
            if (o != null && o.Move.Id == Ms.FUSION_FLARE && b.HasCondition("FusionFlare")) m = 0x2000;
          }
          break;
        case Ms.FACADE: //263
          if (atk.Attacker.State != PokemonState.Normal && atk.Attacker.State != PokemonState.FRZ) m = 0x2000;
          break;
        case Ms.BRINE: //362
          if (der.Hp << 1 <= der.Pokemon.MaxHp) m = 0x2000;
          break;
        case Ms.VENOSHOCK: //474
          if (atk.Attacker.State == PokemonState.PSN && atk.Attacker.State == PokemonState.BadlyPSN) m = 0x2000;
          break;
        case Ms.RETALIATE: //514
          if (atk.Attacker.Field.GetCondition<int>("FaintTurn") == der.Controller.TurnNumber - 1) m = 0x2000;
          break;
      }
      //If move was called using Me First.
      if (atk.HasCondition("MeFirst")) m *= 0x1800;
      //If move is SolarBeam in non-sunny, non-default weather.
      if (def.AtkContext.Move.Id == Ms.SOLARBEAM)
      {
        Weather w = def.Defender.Controller.Weather;
        if (w != Weather.IntenseSunlight && w != Weather.Normal) m *= 0x800;
      }

      return m;
    }
  }
}
